notification: improved API

waht opened the pull request:
# Purpose

This PR improves the notification API. It adds new types (both C and kdbtypes.h). It also adds macros for generating API functions for new types and also adds C99 definitions to kdbtypes.h.

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I added unit tests
- [x] I ran all tests and everything went fine
- [x] affected documentation is fixed
- [x] I added code comments, logging, and assertions (see doc/CODING.md)
- [x] release notes are updated (doc/news/_preparation_next_release.md)
- [x] rebase when dbus PR (#1825) is merged into `master`.
- [x] update notification tutorial: explain emergent misbehavior & add guidelines
- [x] update architecture diagram

@markus2330 please review my pull request

waht commented:
jenkins build multiconfig-gcc-stable please

waht commented:
jenkins build multiconfig-gcc47-cmake-options please

waht commented:
jenkins build all please

markus2330 commented:
Contains commits from #1825 please rebase. I could not do a review because the shown diff is huge.

Please also fix the leftovers of #1825 here.

waht commented:
Okay, I've fixed the leftovers and rebased the PR - the diff is much smaller now :)

markus2330 commented on src/include/kdbtypes.h:
```diff
@@ -48,8 +48,44 @@ using octet_t = uint8_t; // default: 0
+typedef uint32_t kdb_unsigned_long_t;
+typedef uint64_t kdb_unsigned_long_long_t;
+
+#if SIZEOF_LONG == 4
```
The `#if` is not necessary. In C99 inttypes.h defines macros for the printf modifiers, e.g., PRIu8

waht replied:
Great, I didn't know about these!

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -153,6 +192,31 @@ static KeyRegistration * elektraInternalnotificationAddNewRegistration (PluginSt
+		if (checkKeyIsBelowOrSame (check, current))
+		{
+			result = 1;
+			break;
```
`return 1;`

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -153,6 +192,31 @@ static KeyRegistration * elektraInternalnotificationAddNewRegistration (PluginSt
+			break;
+		}
+	}
+	return result;
```
`return 0;`

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -170,167 +234,136 @@ void elektraInternalnotificationUpdateRegisteredKeys (Plugin * plugin, KeySet *
+			if (key == NULL)
+			{
+				registeredKey = registeredKey->next;
+				continue;
```
some refactoring here?

markus2330 commented on src/plugins/internalnotification/internalnotification.c:
```diff
@@ -170,167 +234,136 @@ void elektraInternalnotificationUpdateRegisteredKeys (Plugin * plugin, KeySet *
 }
 
+// Generate register and conversion functions
+INTERNALNOTIFICATION_TYPE (int, long int, Int, (strtol (string, &end, 10)),
```
Very compact, well done!

Is this somehow reusable for other places? (maybe as [super macro](http://s11n.net/papers/supermacros_cpp.pdf)?)

markus2330 commented on src/plugins/internalnotification/testmod_internalnotification.c:
```diff
@@ -382,6 +629,37 @@ int main (int argc, char ** argv)
+	test_doUpdateShouldNotUpdateUnregisteredKey ();
+	test_doUpdateShouldUpdateKeyAbove ();
+
 	printf ("\ntestmod_internalnotification RESULTS: %d test(s) done. %d error(s).\n", nbTest, nbError);
```
should use the print from the test suite.

waht commented on src/include/kdbnotificationinternal.h:
```diff
@@ -23,30 +23,61 @@ namespace ckdb
+#define ELEKTRA_NOTIFICATION_TYPE_DEFINITION(TYPE, TYPE_NAME)                                                                              \
+	ELEKTRA_NOTIFICATION_REGISTER_SIGNATURE (TYPE, TYPE_NAME)                                                                          \
+	{                                                                                                                                  \
+		if (!kdb || !key || !variable)                                                                                             \
```
Yes, but I did not use it because this function is user facing.

markus2330 commented on src/plugins/internalnotification/macros/addType.h:
```diff
@@ -0,0 +1,71 @@
+#ifndef TYPE
```
Please all-lowercase filenames.

Please add some docu how to use the supermacro.

Will it work for @domhof (#1605) and to replace support.setfuncname in src/tools/gen/template/template.h?

waht replied:
While my macros were inspired by his PR I think the only reusable part would be the conversion function. If needed we can extract it but details like error handling would need to be discussed. Theoretically we could have a lib dedicated to string to value and value to string conversions.

markus2330 replied:
Yes, let us discuss this.

waht replied:
Ok! I'll think about error handling and reply not later than Thursday.

markus2330 replied:
> undefined in case of an error.

If possible, the value should stay unchanged.

> What should be name of the conversion library?

If you can implement it as (super)macros no extra library is needed. Otherwise elektra-type might be suitable. We should also add `keySetLong`, `keyGetLong`, etc. functions.

waht replied:
> If possible, the value should stay unchanged.

Ok its better to have additional variable declarations & logic in the macro instead of scattering it in the code base.

> If you can implement it as (super)macros no extra library is needed.

Of course! :) A naming macro as input would be required though to avoid name collisions.

> We should also add `keySetLong`, `keyGetLong`, etc. functions.

For such functions a seperate library would be better suited. Also I think this would broaden the scope of the PR too much. For now I'll better stick to the conversion macros.

waht replied:
I added a supermacro called `type_create_to_value` for creating "to value" conversion functions in "include/macros". I decided to return the error and have a pointer for the variable because it improves readability:

```C
int variable = 0;
if (!myIntConversion (key, &variable))
{
    // error, variable is still 0
}
```

is better than
```C
int variable = 0;
int error = 0;
variable = myIntConversion (key, error)
if (error)
{
    // error, variable was changed to 0 because of return
}
```

markus2330 replied:
>  For such functions a seperate library would be better suited. Also I think this would broaden the scope of the PR too much. For now I'll better stick to the conversion macros.

Yes, of course! I am surprised that you put so much effort in supporting all types. I only mentioned keySetLong to make clear what elektra-type could be used for. But header-only without any lib is the better solution for this case.

waht replied:
Ok, great! Then the PR should be completed :)

markus2330 commented on src/plugins/internalnotification/macros/addType.h:
```diff
@@ -0,0 +1,71 @@
+#define CHECK_CONVERSION 1
+#endif
+
+#define CONCAT(X, Y) CONCAT2 (X, Y)
```
Is this needed as macro?

waht replied:
Yes. It concatenates two tokens. It was not necessary before since it was already encapsulated in a macro before expansion. I'm not sure how to describe, so here is an example :)

```C
// in supermacro.h
#define CONCAT(X, Y) CONCAT2 (X, Y)
#define CONCAT2(X, Y) X##Y
#define FUNC_NAME(TYPE_NAME) CONCAT (internalnotificationRegister, TYPE_NAME)
#define FUNC_NAME2(TYPE_NAME) internalnotificationRegister##TYPE_NAME
#define NESTED(TYPE_NAME) FUNC_NAME2 (TYPE_NAME)

FUNC_NAME (TYPE_NAME)
FUNC_NAME2 (TYPE_NAME)
NESTED (TYPE_NAME)

// in some.c
#define TYPE_NAME Int
#include "macros/supermacro.h"
// => gcc -E some.c
internalnotificationRegisterInt
internalnotificationRegisterTYPE_NAME
internalnotificationRegisterInt
```

Before the "supermacro", the `NESTED` macro was the macro that generated functions. A solution would be to add another macro (e.g. `#define FUNC_SIGNATURE(TYPE_NAME) void FUNC_NAME(TYPE_NAME) (void)` but the `CONCAT` macro is more versatile.

markus2330 replied:
Okay, then we should also add `ELEKTRA_CONCAT` next to `ELEKTRA_STRINGIFY`.

markus2330 commented on src/plugins/internalnotification/macros/addType.h:
```diff
@@ -0,0 +1,71 @@
+	}
+	else
+	{
+		ELEKTRA_LOG_WARNING ("conversion failed! string=%s, stopped=%c errno=%d", keyString (key), *end, errno);
```
So the value is silently (unless the logger is enabled) not changed?

Not so ideal...

markus2330 replied:
> (e.g. the type plugin) to allow only valid values the key database.

Yes, that is the goal. But there is no type validation for C types.

> elektraNotificationSetConversionFailedCallback

It could be useful for applications to log and abort. But it is much better if such situations cannnot occur :-)

markus2330 commented:
Sorry, it seems like I created a merge conflict. Please rebase.

waht commented:
jenkins build all please

waht commented:
jenkins build gcc-configure-debian-debug please

waht commented:
jenkins build mingw64 please

waht commented:
The PR is ready for review. Is it safe to rebase now?

markus2330 commented:
Yes, release is done, so it is now a good time for rebase!

waht commented:
jenkins build gcc-configure-debian-optimizations please

waht commented:
jenkins build multiconfig-gcc-stable please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

ingwinlu commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
jenkins build gcc-configure-debian-optimizations please

waht commented:
jenkins build all please

waht commented:
jenkins build all please

waht commented:
jenkins build all please

waht commented:
@markus2330 I added the "reload" example and updated the tutorial, please review.
This PR should be merged before #1835 due to the kdbio-header reorganisation.

PS: Is the "build all" job supposed to be only "pr-merge" and "configure-debian-musl"?

markus2330 commented:
Yes, @ingwinlu nearly is finished with the migration to move all build jobs into one. Seems like musl is the last of the old ones.

markus2330 commented on src/include/kdbio/ev.h:
```diff
@@ -8,7 +8,7 @@
 #define KDB_IOWRAPPER_EV_H_
 
-#include "kdbio.h"
+#include "../kdbio.h"
```
Better to manipulate the include path and include kdbio/kdbio.h here.
I would even keep the kdbio.h top-level so that `<kdbio.h>` can be included.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -257,6 +257,116 @@ int main (void)
 
+### How-To: Reload KDB when Elektra's configuration has changed
+
+This section shows how the notification feature can be used to reload an
```
is used to

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -257,6 +257,116 @@ int main (void)
+**Step 1: Register for changes to Elektra's configuration**
+
+To achieve reloading on Elektra configuration changes we register for changes
+below the key `system/elektra` using
```
Does `/elektra` works too? It might be necessary once we have #1074

waht replied:
Yes, `/elektra` works too. I'll update the example.

markus2330 commented on doc/tutorials/notifications.md:
````diff
@@ -257,6 +257,116 @@ int main (void)
+up and reinitializates KDB.
+
+```C
+void initKdb (void)
````
`KDB ** kdb` as argument?

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -257,6 +257,116 @@ int main (void)
+	elektraIoSetBinding (kdb, binding);
+	elektraNotificationOpen (kdb);
+
+  // registration code from snippet before
```
indent?

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -257,6 +257,116 @@ int main (void)
+ElektraIoTimerOperation * reload;
+
+// main loop setup (e.g. main())
+reload = elektraIoNewTimerOperation (100, 0, initKdb, NULL);
```
Why 100?

waht replied:
The timer interval is 100 milliseconds; I'll add a comment.

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -257,6 +257,116 @@ int main (void)
+	if (kdb != NULL)
+	{
+		// Cleanup notifications and close KDB
+    elektraNotificationClose (kdb);
```
indent

markus2330 commented on doc/tutorials/notifications.md:
````diff
@@ -257,6 +257,116 @@ int main (void)
+}
+```
+
+By following these three steps any application can react to changes to Elektra's
````
"following" confusing here (usually refers to text afterwards). Better to use `By correct application of these ...`

markus2330 commented on examples/notificationAsync.c:
```diff
@@ -1,7 +1,17 @@
+ *   - io_uv binding
+ *   - Transport plugins (e.g. kdb global-mount dbus announce=once dbusrecv)
+ *
+ * Ideas for this example:
```
Relevant keys for this example:

markus2330 commented on examples/notificationAsync.c:
```diff
@@ -100,7 +110,7 @@ int main (void)
 	if (!result)
 	{
-		printf ("could init notification. aborting\n");
+		printf ("could not init notification. aborting\n");
```
comma

markus2330 commented on examples/notificationPolling.c:
```diff
@@ -1,7 +1,13 @@
+ * @brief Example for notification library which repeatedly reads some keys and
+ * reacts to them; see "example_notification_async" for an example without polling
+ *
+ * Ideas for this example:
```
See above.

markus2330 commented on examples/notificationPolling.c:
```diff
@@ -94,7 +100,7 @@ int main (void)
 	if (!result)
 	{
-		printf ("could init notification. aborting\n");
+		printf ("could not init notification. aborting\n");
```
see above

markus2330 commented on examples/notificationReload.c:
```diff
@@ -0,0 +1,187 @@
+ *   - io_glib binding
+ *   - Transport plugins (e.g. kdb global-mount zeromqsend zeromqrecv && kdb run-hub-zeromq)
+ *
+ * Ideas for this example:
```
see above

markus2330 commented on examples/notificationReload.c:
```diff
@@ -0,0 +1,187 @@
+	int result = elektraNotificationOpen (kdb);
+	if (!result)
+	{
+		printf ("could not init notification. aborting\n");
```
see above

markus2330 commented on examples/notificationReload.c:
```diff
@@ -0,0 +1,187 @@
+	if (kdb == NULL)
+	{
+		printf ("could not open KDB. aborting\n");
+		exit (-1);
```
This would return "status & 0377" to the parent. Why not using exit (1)?

waht replied:
No reason really, except the other examples used "return -1"; I'll use `exit (1)`

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -223,6 +232,20 @@ Thanks to Michael Zronek and Vanessa Kos.
 
 [#1887]: https://github.com/ElektraInitiative/libelektra/issues/1887
 
+## Compatibility
```
I think the "Compatibility" above happened at some wrong merge. @sanssecours any idea how this happened?

sanssecours replied:
I think the duplicated header was added in commit 1b2d0d62. Anyway, pull request [#2066](https://github.com/ElektraInitiative/libelektra/pull/2066/files#diff-47ed151a0df6c926f5b1320d140a5b34L227) should fix the problem.

markus2330 commented on examples/CMakeLists.txt:
```diff
@@ -28,3 +31,74 @@ target_link_elektra (kdbintro elektra-kdb)
+
+	target_link_elektra (${EXAMPLE} elektra-kdb elektra-notification)
+
+	# TODO resolve https://github.com/ElektraInitiative/libelektra/issues/2007
```
please use URL issues.libelektra.org/2007

markus2330 commented on examples/notificationAsync.c:
```diff
@@ -115,13 +135,22 @@ int main (void)
+	printVariable (timer); // "value" was automatically updated
+
+	// This allows us to wake the loop from our signal handler
+#ifdef HAVE_LIBUV1
```
@ingwinlu do we have libuv1 installed somewhere?

waht replied:
libuv 1.x was included in the debian docker images, but afaik they are not used by the build server.
On Travis we only have libuv 0.x available. This PR will add 1.x to the build server.

markus2330 commented on src/include/kdbnotification.h:
```diff
@@ -101,37 +104,95 @@ int elektraNotificationOpen (KDB * kdb);
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (int, Int)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (unsigned int, UnsignedInt)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (long, Long)
+ELEKTRA_NOTIFICATION_TYPE_DECLARATION (unsigned long, UnsignedLong)
```
why not long long?

waht replied:
No problem, I'll add it.

markus2330 commented on src/include/macros/type_create_to_value.h:
```diff
@@ -0,0 +1,98 @@
+ *                          to check additionally for a specified range.
+ * @param DISABLE_UNDEF_PARAMETERS define to disable undefining of parameters after the macro. Use if parameters
+ *                                 are used within another supermacro.
+ */
```
license BSD missing

waht commented:
jenkins build libelektra please

waht commented:
jenkins build libelektra please

waht commented:
@markus2330 I fixed some unsigned conversion issues, the PR is ready for a re-review :)

markus2330 commented on src/plugins/internalnotification/README.md:
```diff
@@ -19,47 +19,9 @@ Application developers should use the
+This plugin exports various functions starting with "register*" below
+`system/elektra/modules/internalnotification/exports/`.
+These functions should not be used directly.
+Instead the notification API should be used.
```
Link to notification API.

markus2330 commented on src/libs/notification/example/README.md:
```diff
@@ -1,28 +0,0 @@
-# Example Applications for Notifications
```
Why is this README removed?

Btw. the  src/libs/notification/example is quite hidden, maybe you can add links to it also at other places.

waht replied:
I forgot to delete the folder - I moved them to /examples. 

About the README: The existing examples didn't have one so I removed it. On second thought I should add instructions about expected plugins to the examples (i.e. do a `printf()`).

waht replied:
Here is an updated example: https://github.com/ElektraInitiative/libelektra/pull/1850/files#diff-98f467ea2a20eaed5714de95cc0d4142R193

markus2330 commented on src/libs/invoke/invoke.c:
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
+/**
+ * Call a deferable function on a plugin handle.
+ * If the function is exported by the plugin it is directly invoked,
+ * if the plugin supports deferring calls, the call is deferred.
```
What if it is exported and also supports deferring calls?

waht replied:
The code currently favors direct calls and does not defer it. I gave it some thought and it would be useful to do both when possible. 
If, for example, the list plugin used I/O bindings and exposed a "setIoBinding" function then it would be useful to still have the call deferred to its encapsulated plugins.

markus2330 commented on src/libs/invoke/invoke.c:
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
+ * If the function is exported by the plugin it is directly invoked,
+ * if the plugin supports deferring calls, the call is deferred.
+ *
+ * The parameters key set can be freed afterwards.
```
Write this directly to the param description.

markus2330 commented on src/libs/invoke/invoke.c:
```diff
@@ -337,31 +338,75 @@ void elektraInvokeClose (ElektraInvokeHandle * handle, Key * errorKey)
+ * @param  elektraPluginFunctionName function name
+ * @param  parameters                parameter key set
+ * @retval 0 on success
+ * @retval 1 when function not exported and deferring is unsupported by plugin
```
should be negative (like other errors). How can the user know which situation happened?

waht replied:
I'm not sure if the user needs to know what happened. I think it is only interesting to know whether the call failed (cannot deferr and cannot call directly) or whether there was success (either deferred or directly called [or both]).

markus2330 commented on src/libs/README.md:
```diff
@@ -82,3 +82,15 @@ data structures.
+
+    libelektra-notification.so
+
+**[notification](notification/)** provides the [notification API](https://doc.libelektra.org/api/current/html/group__kdbnotification.html).
```
Maybe add link to examples also here.

markus2330 commented on src/libs/README.md:
```diff
@@ -82,3 +82,15 @@ data structures.
+
+    libelektra-io.so
+
+**[io](io/)** provides the common [I/O binding API](https://doc.libelektra.org/api/current/html/group__kdbio.html).
```
a very short description would be nice. It is unclear what "I/O" is in this context.

markus2330 commented on examples/notificationReload.c:
```diff
@@ -0,0 +1,190 @@
+GMainLoop * loop;
+
+KDB * kdb;
+Key * parentKey;
```
are all of them needed globally? Otherwise the example is great!

waht replied:
Thank you! No they can also be moved to a struct which is passed as user data - I updated the example to demo usage of the required functions (`elektraIo[Binding|Fd|Timer|Idle]GetData()`).

waht replied:
Please check out the [updated example](https://github.com/ElektraInitiative/libelektra/pull/1850/files#diff-98f467ea2a20eaed5714de95cc0d4142).

markus2330 commented on doc/tutorials/notifications.md:
```diff
@@ -53,125 +47,163 @@ notifications but not whether notifications are sent or which transport is used.
 void main (void)
 {
-	KDB* repo;
+  KDB* repo;
```
Any specific reason why the change to spaces here?

waht replied:
No, no reason. Might be the editor messing with identation or me forgetting about it when pasting code.

waht commented:
jenkins build libelektra please

markus2330 commented:
Looks much better now, is it ready to merge?

markus2330 commented:
@tom-wa would be great to have this for the upcoming release which hopefully happens today.

markus2330 commented:
Sorry, I wanted to ping @waht here.

waht commented:
Sure! I just added a commit improving the connection timeout resolution.

markus2330 commented:
Thank you, finally it is done :smile:
